src/undo.h

```diff
@@ -34,7 +34,7 @@ struct TxInUndoFormatter
 
     template<typename Stream>
     void Unser(Stream &s, Coin& txout) {
-        unsigned int nCode = 0;
+        uint32_t nCode = 0;
```

Comment:
Again, not sure if I understand this correctly.

The undefined behavior happens because standard only says `unsigned int nCode` has to be 16+ bits.

 If `nCode` is 32 bits everything works perfectly. Any serialized VarInt value that is 2**32 or more will trigger "ReadVarInt(): size too large" exception, and any smaller values will be broken up into height and fcoinbase correctly.

If `nCode` is less than 32 bits, ReadVarInt() exceptions will be triggered unnecessarily. If `nCode` is more than 32 bits, top bits that are set will be silently discarded without triggering the ReadVarInt() exceptions that should be triggered.

Chaning `nCode` to be 32 bits instead of having an implementation defined width gives the right behavior on all platforms.

Again, I'm wondering if this is right or something else is going on.

Reply:
> The undefined behavior happens because standard only says unsigned int nCode has to be 16+ bits.

I don't think any undefined behavior has been observed here. This has just been changed for clarity and should also not result in a change in the compiled code.

Reply:
Yes this is mostly for the sake of clarity. I'm struggling to see bitcoin run on platforms where `unsigned int` is different than 32 bits but who knows.

> If nCode is less than 32 bits, ReadVarInt() exceptions will be triggered unnecessarily.

If on an imaginary platform `unsigned int` is 16 bits, then yes `nCode` have to be less than `2**16/(6*24*365)<2 years` so this code would already hit the exception in `ReadVarInt`. 

> If nCode is more than 32 bits, top bits that are set will be silently discarded without triggering the ReadVarInt() exceptions that should be triggered.

Yes. In that case adding the new exception would be necessary (limiting the threshold to `unsigned int`, not `uint32_t`).

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18433#discussion_r399552921

> I don't think any undefined behavior has been observed here. This has just been changed for clarity and should also not result in a change in the compiled code.

Good to know. Any I don't think it was right for me to call it "undefined behavior", more like "platform specific behavior" only on a theoretical platform where an int is 31 bits with when height is somehow >= 2**30.

In the future, it would be good to do fuzzer fixes and code cleanups in separate commits. Or at least say what the different changes are intended to do in the commit message.  The changes are enough work to understand without having to guess which are real and which are no-ops!